From 08b5ebefc9970ce139d7b501d8c3d53176443b13 Mon Sep 17 00:00:00 2001
From: Ido Reis <idor@ti.com>
Date: Mon, 30 Apr 2012 13:52:12 +0300
Subject: [PATCH 10/21] board-4430: Add support for mmc5/dat1 irq source *PRELIMINARY*

Set dat1_gpio and implement a remuxing function for using mmc5/dat1 as an
irq source by omap_hsmmc

Signed-off-by: Ido Yariv <ido@wizery.com>
Signed-off-by: Eliad Peller <eliad@wizery.com>
Signed-off-by: Ido Reis <idor@ti.com>
---
 arch/arm/mach-omap2/board-4430sdp.c |   28 ++++++++++++++++++++++++++++
 1 files changed, 28 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/board-4430sdp.c b/arch/arm/mach-omap2/board-4430sdp.c
index 47f1573..505278c 100644
--- a/arch/arm/mach-omap2/board-4430sdp.c
+++ b/arch/arm/mach-omap2/board-4430sdp.c
@@ -79,6 +79,7 @@
 
 #define GPIO_WIFI_PMENA		54
 #define GPIO_WIFI_IRQ		53
+#define GPIO_MMC5_DAT1		148
 #define OMAP_HDMI_HPD_ADDR	0x4A100098
 #define OMAP_HDMI_PULLTYPE_MASK	0x00000010
 
@@ -486,6 +487,29 @@ static struct omap_musb_board_data musb_board_data = {
 	.power			= 200,
 };
 
+static void remux_mmc5_dat1(bool clocks_enabled)
+{
+	void __iomem *mux_base = NULL;
+	u32 val;
+
+	mux_base = ioremap(0x4A100000, 0x1000);
+	if (!mux_base) {
+		pr_err("%s: Could not remap mux base", __func__);
+		return;
+	}
+
+	val = __raw_readl(mux_base + 0x14C) & 0x0000FFFF;
+	if (clocks_enabled)
+		val |= (OMAP_MUX_MODE0 | OMAP_PIN_INPUT_PULLUP) << 16;
+	else
+		val |= (OMAP_MUX_MODE3 | OMAP_PIN_INPUT_PULLUP |
+			OMAP_PIN_OFF_WAKEUPENABLE) << 16;
+
+	__raw_writel(val, mux_base + 0x14C);
+
+	iounmap(mux_base);
+}
+
 static struct omap2_hsmmc_info mmc[] = {
 	{
 		.mmc		= 2,
@@ -493,6 +517,7 @@ static struct omap2_hsmmc_info mmc[] = {
 					MMC_CAP_1_8V_DDR,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= -EINVAL,
 		.nonremovable   = true,
 		.ocr_mask	= MMC_VDD_29_30,
 		.no_off_init	= true,
@@ -502,6 +527,7 @@ static struct omap2_hsmmc_info mmc[] = {
 		.caps		= MMC_CAP_4_BIT_DATA | MMC_CAP_8_BIT_DATA |
 					MMC_CAP_1_8V_DDR,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= -EINVAL,
 	},
 	{
 		.mmc		= 5,
@@ -509,8 +535,10 @@ static struct omap2_hsmmc_info mmc[] = {
 				  MMC_CAP_SDIO_IRQ | MMC_CAP_ASYNC_SDIO_IRQ,
 		.gpio_cd	= -EINVAL,
 		.gpio_wp	= -EINVAL,
+		.gpio_dat1	= GPIO_MMC5_DAT1,
 		.ocr_mask	= MMC_VDD_165_195,
 		.nonremovable	= true,
+		.remux_dat1	= remux_mmc5_dat1,
 	},
 	{}	/* Terminator */
 };
-- 
1.7.0.4

